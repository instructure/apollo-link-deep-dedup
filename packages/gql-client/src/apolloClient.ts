import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloClient } from 'apollo-client';
import { ApolloLink } from 'apollo-link';
import { createHttpLink } from 'apollo-link-http';
import fetch from 'cross-fetch';

import { DeepDedupLink } from '../../apollo-link-deep-dedup/lib/index';

const SERVER_URI = 'http://localhost:3000';
const DATA_API_URI = `${SERVER_URI}/graphql`;

// Initialize client cache
const cache = new InMemoryCache();

// Initialize Apollo Links
const deepDedupLink = new DeepDedupLink({
    cache: cache,
});

const httpLink = createHttpLink({
    uri: DATA_API_URI,
    fetch: fetch,
});

const link = ApolloLink.from([
    deepDedupLink,
    httpLink,
]);

// Initialize Apollo Client
export const client = new ApolloClient({
    cache: cache,
    link, // [apollo-deep-dedup-link, apollo-http-link]
});
