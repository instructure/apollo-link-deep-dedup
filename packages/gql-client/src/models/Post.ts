export default interface Post {
    id: number;
    authorId: number;
    title: string;
    votes: number;
}
