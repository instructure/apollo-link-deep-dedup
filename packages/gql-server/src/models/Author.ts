import Post from './Post';

export default interface Author {
    id: number;
    firstName: string;
    lastName: string;
}
