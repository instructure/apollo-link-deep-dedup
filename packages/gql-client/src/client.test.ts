import 'jest';

import { client } from './apolloClient';
import {
    fetchAllAuthors,
    fetchAllPosts,
    fetchAuthorById,
    upvotePostByPostId,
} from './queries';

import Author from './models/Author';
import Post from './models/Post';

describe('Client', () => {
    it('fetches all authors', async () => {
        const result = await client.query(fetchAllAuthors());
        const authors: Author[] = result.data as Author[];
        expect(authors).toMatchSnapshot();
    });

    it('fetches all posts', async () => {
        const result = await client.query(fetchAllPosts());
        const posts: Post[] = result.data as Post[];
        expect(posts).toMatchSnapshot();
    });

    it('fetches author by id', async () => {
        const authorId = 1;
        const result = await client.query(fetchAuthorById(authorId));
        const author = result.data as Author;
        expect(author).toMatchSnapshot();
    });

    it('upvotes post by post id', async () => {
        const postId = 1;
        const result = await client.mutate(upvotePostByPostId(postId));
        const upvotedPost: Post = result.data.upvotePost as Post;
        expect(upvotedPost).toHaveProperty('id', postId);
        expect(upvotedPost.votes).toBeGreaterThan(1);
    });
});

