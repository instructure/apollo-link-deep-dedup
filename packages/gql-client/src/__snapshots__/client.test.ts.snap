// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client fetches all authors 1`] = `
Object {
  "authors": Array [
    Object {
      "__typename": "Author",
      "firstName": "Tom",
      "id": 1,
      "lastName": "Coleman",
      "posts": Array [
        Object {
          "__typename": "Post",
          "id": 1,
          Symbol(id): "Post:1",
        },
      ],
      Symbol(id): "Author:1",
    },
    Object {
      "__typename": "Author",
      "firstName": "Sashko",
      "id": 2,
      "lastName": "Stubailo",
      "posts": Array [
        Object {
          "__typename": "Post",
          "id": 2,
          Symbol(id): "Post:2",
        },
        Object {
          "__typename": "Post",
          "id": 3,
          Symbol(id): "Post:3",
        },
      ],
      Symbol(id): "Author:2",
    },
    Object {
      "__typename": "Author",
      "firstName": "Mikhail",
      "id": 3,
      "lastName": "Novikov",
      "posts": Array [
        Object {
          "__typename": "Post",
          "id": 4,
          Symbol(id): "Post:4",
        },
      ],
      Symbol(id): "Author:3",
    },
  ],
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`Client fetches all posts 1`] = `
Object {
  "posts": Array [
    Object {
      "__typename": "Post",
      "author": Object {
        "__typename": "Author",
        "id": 1,
        Symbol(id): "Author:1",
      },
      "id": 1,
      "title": "Introduction to GraphQL",
      Symbol(id): "Post:1",
    },
    Object {
      "__typename": "Post",
      "author": Object {
        "__typename": "Author",
        "id": 2,
        Symbol(id): "Author:2",
      },
      "id": 2,
      "title": "Welcome to Meteor",
      Symbol(id): "Post:2",
    },
    Object {
      "__typename": "Post",
      "author": Object {
        "__typename": "Author",
        "id": 2,
        Symbol(id): "Author:2",
      },
      "id": 3,
      "title": "Advanced GraphQL",
      Symbol(id): "Post:3",
    },
    Object {
      "__typename": "Post",
      "author": Object {
        "__typename": "Author",
        "id": 3,
        Symbol(id): "Author:3",
      },
      "id": 4,
      "title": "Launchpad is Cool",
      Symbol(id): "Post:4",
    },
  ],
  Symbol(id): "ROOT_QUERY",
}
`;

exports[`Client fetches author by id 1`] = `
Object {
  "author": Object {
    "__typename": "Author",
    "firstName": "Tom",
    "lastName": "Coleman",
    Symbol(id): "$ROOT_QUERY.author({\\"id\\":1})",
  },
  Symbol(id): "ROOT_QUERY",
}
`;
